### PR TITLE
 bpo-25041: Document AF_PACKET socket address format 

### PR DESCRIPTION
**Pull request by csabella**
Add documentation for AF_PACKET. 

Includes suggestions by @berkerpeksag  and @vadmium  from the ticket.

https://bugs.python.org/issue25041

### CODE REVIEW

File: Doc/library/socket.rst [hunk 1]
```diff
@@ -161,10 +161,25 @@ created.  Socket addresses are represented as follows:
 
   .. versionadded:: 3.7
 
-- Certain other address families (:const:`AF_PACKET`, :const:`AF_CAN`)
-  support specific representations.
-
-  .. XXX document them!
+- The :const:`AF_PACKET` address family is a Linux-only packet address protocol
+  represented by the tuple ``(ifname, proto[, pkttype[, hatype[, addr]]])``
```
Comment (vadmium) [inline Doc/library/socket.rst]: I haven’t heard of “packet address protocols” before. What does it mean, and why did you add it?

**Comment by vadmium** [comment]
Apart from the strange wording introducing the address tuple, the rest looks okay to me.

File: Doc/library/socket.rst (hunk 1 above)
Comment (csabella) [reply Doc/library/socket.rst]: The man page for [socket](http://man7.org/linux/man-pages/man2/socket.2.html) has:
```
       Name                Purpose                          Man page
       AF_UNIX, AF_LOCAL   Local communication              unix(7)
       AF_INET             IPv4 Internet protocols          ip(7)
       AF_INET6            IPv6 Internet protocols          ipv6(7)
       AF_IPX              IPX - Novell protocols
       AF_NETLINK          Kernel user interface device     netlink(7)
       AF_X25              ITU-T X.25 / ISO-8208 protocol   x25(7)
       AF_AX25             Amateur radio AX.25 protocol
       AF_ATMPVC           Access to raw ATM PVCs
       AF_APPLETALK        AppleTalk                        ddp(7)
       AF_PACKET           Low level packet interface       packet(7)
       AF_ALG              Interface to kernel crypto API
```
I think I should have said `Linux-only packet interface` or `Linux-only low level packet interface`.  I'm not sure why I changed that to protocol.  I may have read it in a description somewhere, but I can't find the source now.

File: Doc/library/socket.rst [hunk 2]
```diff
@@ -161,10 +161,25 @@ created.  Socket addresses are represented as follows:
...
+  where:
+
+  - *ifname* - String specifying the device name.
+  - *proto* - An endian-corrected integer specifying the Ethernet protocol
```
Comment (benjaminp) [inline Doc/library/socket.rst]: I would say "in network byte order" rather than "endian-corrected".

File: Doc/library/socket.rst [hunk 3]
```diff
@@ -161,10 +161,25 @@ created.  Socket addresses are represented as follows:
...
+    number.
+  - *pkttype* - Optional constant integer specifying the packet type:
```
Comment (benjaminp) [inline Doc/library/socket.rst]: Remove "constant".

File: Doc/library/socket.rst [hunk 4]
```diff
@@ -363,6 +378,16 @@ Constants
    .. versionadded:: 3.7
 
 
+.. data:: AF_PACKET
+          PF_PACKET
+          PACKET_*
+
+   Many constants of these forms, documented in the Linux documentation, are
```
Comment (benjaminp) [inline Doc/library/socket.rst:398]: "Linux documentation" -> "man pages"

File: Doc/library/socket.rst (hunk 1 above)
Comment (benjaminp) [reply Doc/library/socket.rst]: I think the important thing here is to describe that it's a low-level interface directly to network devices.

**Review by benjaminp — CHANGES_REQUESTED** [review]

File: Doc/library/socket.rst (hunk 4 above)
Comment (csabella) [reply Doc/library/socket.rst:398]: "Linux documentation" is used in the other definitions within this section, so I had used it for consistency.  Should I change them all?

Comment (benjaminp) [reply Doc/library/socket.rst:398]: Ah, sorry, I didn't realize that term was already in use here. I suppose this is fine to leave as is.

File: Doc/library/socket.rst (hunk 1 above)
Comment (csabella) [reply Doc/library/socket.rst]: Done.

File: Doc/library/socket.rst (hunk 2 above)
Comment (csabella) [reply Doc/library/socket.rst]: Done.

File: Doc/library/socket.rst (hunk 3 above)
Comment (csabella) [reply Doc/library/socket.rst]: Done.

File: Doc/library/socket.rst (hunk 4 above)
Comment (csabella) [reply Doc/library/socket.rst:398]: Looking at git blame, it seems that one person started it and then all the others continued it, presumably for consistency.  If `man pages` is better, then maybe it would be best to change it now instead of propagating `Linux documentation` every time?

Comment (benjaminp) [reply Doc/library/socket.rst:398]: I think it would make sense to change that, but that should probably be another change.
